{
	"schema_version": "1.4.0",
	"id": "GHSA-5h86-8mv2-jq9f",
	"modified": "2024-01-29T22:31:03Z",
	"published": "2024-01-29T22:31:03Z",
	"aliases": [
		"CVE-2024-23334"
	],
	"summary": "aiohttp is vulnerable to directory traversal",
	"details": "### Summary\nImproperly configuring static resource resolution in aiohttp when used as a web server can result in the unauthorized reading of arbitrary files on the system.\n\n### Details\nWhen using aiohttp as a web server and configuring static routes, it is necessary to specify the root path for static files. Additionally, the option 'follow_symlinks' can be used to determine whether to follow symbolic links outside the static root directory. When 'follow_symlinks' is set to True, there is no validation to check if a given file path is within the root directory.This can lead to directory traversal vulnerabilities, resulting in unauthorized access to arbitrary files on the system, even when symlinks are not present.\n\ni.e. An application is only vulnerable with setup code like:\n```\napp.router.add_routes([\n    web.static(\"/static\", \"static/\", follow_symlinks=True),  # Remove follow_symlinks to avoid the vulnerability\n])\n```\n\n### Impact\nThis is a directory traversal vulnerability with CWE ID 22. When using aiohttp as a web server and enabling static resource resolution with `follow_symlinks` set to True, it can lead to this vulnerability. This vulnerability has been present since the introduction of the `follow_symlinks` parameter.\n\n### Workaround\nEven if upgrading to a patched version of aiohttp, we recommend following these steps regardless.\n\nIf using `follow_symlinks=True` outside of a restricted local development environment, disable the option immediately. This option is NOT needed to follow symlinks which point to a location _within_ the static root directory, it is _only_ intended to allow a symlink to break out of the static directory. Even with this CVE fixed, there is still a substantial risk of misconfiguration when using this option on a server that accepts requests from remote users.\n\nAdditionally, aiohttp has always recommended using a reverse proxy server (such as nginx) to handle static resources and _not_ to use these static resources in aiohttp for production environments. Doing so also protects against this vulnerability, and is why we expect the number of affected users to be very low.\n\n-----\n\nPatch: https://github.com/aio-libs/aiohttp/pull/8079/files",
	"affected": [
		{
			"package": {
				"ecosystem": "PyPI",
				"name": "aiohttp"
			},
			"ranges": [
				{
					"type": "ECOSYSTEM",
					"events": [
						{
							"introduced": "1.0.5"
						},
						{
							"fixed": "3.9.2"
						}
					]
				}
			]
		}
	],
	"references": [
		{
			"type": "WEB",
			"url": "https://github.com/aio-libs/aiohttp/security/advisories/GHSA-5h86-8mv2-jq9f"
		},
		{
			"type": "WEB",
			"url": "https://github.com/aio-libs/aiohttp/pull/8079/files"
		},
		{
			"type": "WEB",
			"url": "https://github.com/aio-libs/aiohttp/commit/1c335944d6a8b1298baf179b7c0b3069f10c514b"
		},
		{
			"type": "PACKAGE",
			"url": "https://github.com/aio-libs/aiohttp"
		}
	],
	"database_specific": {
		"cwe_ids": [
			"CWE-22"
		],
		"github_reviewed": true,
		"github_reviewed_at": "2024-01-29T22:31:03Z",
		"nvd_published_at": null,
		"severity": "MODERATE"
	}
}
